polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -29,297 +32,548 @@ namespace {
+  for (auto it : llvm::zip(loop.getResults(), ifOp.getResults()))
+    std::get<0>(it).replaceUsesWithIf(std::get<1>(it), [&](OpOperand &op) {
+      Block *useBlock = op.getOwner()->getBlock();
+      return useBlock != ifOp.thenBlock();
+    });
 }
 
-void moveParallelLoopInvariantCode(scf::ParallelOp loop) {
-  SmallVector<Operation *, 8> opsToMove;
-  collectHoistableOperation(loop, opsToMove);
-  if (opsToMove.empty())
-    return;
+//===----------------------------------------------------------------------===//
+// SCFParallelLoopGuardBuilder
+//===----------------------------------------------------------------------===//
+
+void SCFParallelGuardBuilder::guardLoop() const {
+  scf::IfOp ifOp = createGuard();
+  loop->moveBefore(ifOp.thenYield());
+  replaceUsesOfLoopReturnValues(ifOp);
+}
 
-  // Move all operations we found to be invariant outside of the loop.
-  OpBuilder b(loop);
+Value SCFParallelGuardBuilder::createGuardExpr() const {
   Value cond;
   for (auto pair :
        llvm::zip(loop.getLowerBound(), loop.getUpperBound(), loop.getStep())) {
-    const Value val = b.create<arith::CmpIOp>(
+    const Value val = builder.create<arith::CmpIOp>(
         loop.getLoc(), arith::CmpIPredicate::sle,
-        b.create<arith::AddIOp>(loop.getLoc(), std::get<0>(pair),
-                                std::get<2>(pair)),
+        builder.create<arith::AddIOp>(loop.getLoc(), std::get<0>(pair),
+                                      std::get<2>(pair)),
         std::get<1>(pair));
     cond = cond ? static_cast<Value>(
-                      b.create<arith::AndIOp>(loop.getLoc(), cond, val))
+                      builder.create<arith::AndIOp>(loop.getLoc(), cond, val))
                 : val;
   }
+  return cond;
+}
 
-  auto ifOp = b.create<scf::IfOp>(loop.getLoc(), TypeRange(), cond);
-  loop->moveBefore(ifOp.thenYield());
+scf::IfOp SCFParallelGuardBuilder::createGuard() const {
+  return builder.create<scf::IfOp>(loop.getLoc(), TypeRange(),
+                                   createGuardExpr());
+}
 
-  for (Operation *op : opsToMove)
-    loop.moveOutOfLoop(op);
+void SCFParallelGuardBuilder::replaceUsesOfLoopReturnValues(
+    scf::IfOp &ifOp) const {
+  // TODO
+}
 
-  LLVM_DEBUG({
-    loop.print(llvm::dbgs() << "\nModified loop:\n");
-    llvm::dbgs() << "\n----------------\n";
-  });
+//===----------------------------------------------------------------------===//
+// AffineForLoopGuardBuilder
+//===----------------------------------------------------------------------===//
+
+void AffineForGuardBuilder::guardLoop() const {
+  AffineIfOp ifOp = createGuard();
+
+  bool yieldsResults = !loop.getResults().empty();
+  if (yieldsResults) {
+    ifOp.getThenBodyBuilder().create<AffineYieldOp>(loop.getLoc(),
+                                                    loop.getResults());
+    ifOp.getElseBodyBuilder().create<AffineYieldOp>(loop.getLoc(),
+                                                    loop.getIterOperands());
+  }
```

Comment:
maybe better to be placed in `createGuard`

Reply:
next PR in this series  will address this comment.